### PR TITLE
Added axis parameter to tf.losses.softmax_cross_entropy. Fixes #20866

### LINKED ISSUE #20866: Why tf.losses.softmax_cross_entropy doesn't have "dim" (axis) argument?
Hi!
The function "tf.losses.softmax_cross_entropy" calls "nn.softmax_cross_entropy_with_logits_v2" which has a dim argument with a default value of -1.
Can we add the same "dim" or "axis" parameter to the calling function " tf.losses.softmax_cross_entropy", so that we can choose which axis to apply the softmax  ?
Thanks!

### PR DESCRIPTION
**Pull request by coder3101**
The internal implementation of tf.losses.softmax_cross_entropy, uses nn.softmax_cross_entropy_with_logits_v2 which has a dim argument with a default value of -1. The new argument axis=-1 wraps the dim argument of nn.softmax_cross_entropy_with_logits_v2. Documentations also updated.
This will fix #20866

### CODE REVIEW

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aaroey: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aaroey: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aaroey: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aaroey: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aaroey: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aaroey: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

File: tensorflow/python/ops/losses/losses_impl.py [hunk 1]
```diff
@@ -762,6 +762,7 @@ def softmax_cross_entropy(
   Args:
     onehot_labels: One-hot-encoded labels.
     logits: Logits outputs of the network.
+    axis: Axis for `num_classes`. Defaulted to -1, i.e Last dimension.
```
Comment (josh11b) [inline tensorflow/python/ops/losses/losses_impl.py]: Defaulted -> Defaults

File: tensorflow/python/ops/losses/losses_impl.py [hunk 2]
```diff
@@ -739,7 +739,7 @@ def sigmoid_cross_entropy(
 
 @tf_export("losses.softmax_cross_entropy")
 def softmax_cross_entropy(
-    onehot_labels, logits, weights=1.0, label_smoothing=0, scope=None,
+    onehot_labels, logits, axis=-1, weights=1.0, label_smoothing=0, scope=None,
```
Comment (josh11b) [inline tensorflow/python/ops/losses/losses_impl.py]: In general it is not backward compatible to add new arguments anywhere except the end. If you think this is a better position for this argument, you can create two versions. A v1 version with it added at the end, a v2 version with it in this position, and a rule in tensorflow/tools/compatibility/tf_upgrade_v2.py.

**Review by josh11b — COMMENTED** [review]
Sorry for the delay in this review!

File: tensorflow/python/ops/losses/losses_impl.py (hunk 1 above)
Comment (coder3101) [reply tensorflow/python/ops/losses/losses_impl.py]: Okay This change has been made

File: tensorflow/python/ops/losses/losses_impl.py (hunk 2 above)
Comment (coder3101) [reply tensorflow/python/ops/losses/losses_impl.py]: Done please review the proposed changed.

**Comment by coder3101** [comment]
I had to force push because I did a rebase to upstream. Hope that doesn't cause any issues. I am running the `bazel build tensorflow/tools/compatibility/update:generate_v2_reorders_map` once done i will push the changes.

**Comment by coder3101** [comment]
I still have an issue running `bazel-bin/tensorflow/tools/compatibility/update/generate_v2_reorders_map` 

It not only fails on my run but even the tensorflow/master but it fails on tensorflow/master as well. (I tried out of curiosity) with the following error :

```
Traceback (most recent call last):
  File "/home/ashar786khan/tf2/tensorflow/bazel-bin/tensorfl
ow/tools/compatibility/update/generate_v2_reorders_map.runfi
les/org_tensorflow/tensorflow/tools/compatibility/update/gen
erate_v2_reorders_map.py", line 141, in <module>
    app.run(main=main)
  File "/home/ashar786khan/tf2/tensorflow/bazel-bin/tensorfl
ow/tools/compatibility/update/generate_v2_reorders_map.runfi
les/org_tensorflow/tensorflow/python/platform/app.py", line
40, in run
    _run(main=main, argv=argv, flags_parser=_parse_flags_tol
erate_undef)
  File "/home/ashar786khan/tf2/tensorflow/bazel-bin/tensorfl
ow/tools/compatibility/update/generate_v2_reorders_map.runfi
les/absl_py/absl/app.py", line 300, in run
    _run_main(main, args)
  File "/home/ashar786khan/tf2/tensorflow/bazel-bin/tensorfl
ow/tools/compatibility/update/generate_v2_reorders_map.runfi
les/absl_py/absl/app.py", line 251, in _run_main
    sys.exit(main(argv))
  File "/home/ashar786khan/tf2/tensorflow/bazel-bin/tensorfl
ow/tools/compatibility/update/generate_v2_reorders_map.runfi
les/org_tensorflow/tensorflow/tools/compatibility/update/gen
erate_v2_reorders_map.py", line 137, in main
    update_reorders_v2(_OUTPUT_FILE_PATH)
  File "/home/ashar786khan/tf2/tensorflow/bazel-bin/tensorfl
ow/tools/compatibility/update/generate_v2_reorders_map.runfi
les/org_tensorflow/tensorflow/tools/compatibility/update/gen
erate_v2_reorders_map.py", line 133, in update_reorders_v2
    file_io.write_string_to_file(output_file_path, renames_f
ile_text)
  File "/home/ashar786khan/tf2/tensorflow/bazel-bin/tensorfl
ow/tools/compatibility/update/generate_v2_reorders_map.runfi
les/org_tensorflow/tensorflow/python/lib/io/file_io.py", lin
e 351, in write_string_to_file
    f.write(file_content)
  File "/home/ashar786khan/tf2/tensorflow/bazel-bin/tensorfl
ow/tools/compatibility/update/generate_v2_reorders_map.runfi
les/org_tensorflow/tensorflow/python/lib/io/file_io.py", lin
e 108, in write
    self._prewrite_check()
File "/home/ashar786khan/tf2/tensorflow/bazel-bin/tensorfl
ow/tools/compatibility/update/generate_v2_reorders_map.runfi
les/org_tensorflow/tensorflow/python/lib/io/file_io.py", lin
e 94, in _prewrite_check
    compat.as_bytes(self.__name), compat.as_bytes(self.__mod
e), status)
  File "/home/ashar786khan/tf2/tensorflow/bazel-bin/tensorfl
ow/tools/compatibility/update/generate_v2_reorders_map.runfi
les/org_tensorflow/tensorflow/python/framework/errors_impl.p
y", line 548, in __exit__
    c_api.TF_GetCode(self.status.status))
tensorflow.python.framework.errors_impl.NotFoundError: third
_party/tensorflow/tools/compatibility/reorders_v2.py; No suc
h file or directory

```

**Comment by rthadur** [comment]
@coder3101 can you please resolve conflicts

**Comment by coder3101** [comment]
@rthadur Fixed

**Comment by annarev** [comment]
This change looks fine to me. But my understanding, it might need to be cherrypicked into 1.14 at this point since it changes API (adds an argument in the middle of function signature).
@josh11b @martinwicke Should we accept and cherrypick this change or instead ask author to add `axis` argument at the end?

**Comment by annarev** [comment]
My reply above was incorrect.
Actually, all the current losses under `tf.losses` are deprecated in 2.0. Instead, They are replaced by losses from `tf.keras.losses`.

`axis` would need to be added here instead: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/losses.py#L405

**Pull request closed without merging**